### PR TITLE
Flow-Direction Normalization: rotate coords by -AoA to streamwise frame

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Rotate the input (x, y) coordinates by -AoA so the x-axis is always the **streamwise direction** and y is always cross-stream. This makes input coordinate statistics AoA-independent at the normalization level, which should improve OOD generalization (p_oodc, p_re).

**Key difference from SE(2) canonicalization (#2270, FAILED):** That experiment used a chord-aligned frame requiring LE/TE detection, which caused (a) stats mismatch with precomputed normalization and (b) DSDF gradient inconsistency. THIS approach uses a KNOWN rotation angle (AoA, always available as a scalar) and does NOT rotate DSDF features — only raw (x, y) coordinates and velocity targets (Ux, Uy). Pressure is scalar, unchanged.

**Physics motivation:** The Navier-Stokes equations are naturally decomposed into streamwise and cross-stream components. Aligning coordinates with the flow removes AoA from the coordinate statistics, making every sample look like zero-AoA at the coordinate level while preserving all other features unchanged.

## Instructions

Apply a 2D rotation by -AoA to (x, y) input coordinates and (Ux, Uy) output targets at the beginning and end of the forward pass.

### Implementation

1. **Read `prepare_multi.py`** to identify which input channels are (x, y) coordinates and which are DSDF/geometric features. Critical: only rotate raw coordinates, NOT DSDF channels.

2. **Apply rotation to inputs:**
```python
aoa_rad = batch_aoa  # [B] per-sample AoA in radians
cos_a = torch.cos(-aoa_rad)  # rotate by -AoA
sin_a = torch.sin(-aoa_rad)

# Rotate only (x, y) coordinate channels:
x_coord = x[:, :, 0].clone()  # channel 0 = x
y_coord = x[:, :, 1].clone()  # channel 1 = y
x[:, :, 0] = cos_a.unsqueeze(1) * x_coord - sin_a.unsqueeze(1) * y_coord
x[:, :, 1] = sin_a.unsqueeze(1) * x_coord + cos_a.unsqueeze(1) * y_coord
```

3. **Apply rotation to velocity targets:**
```python
Ux = target[:, :, 0].clone()
Uy = target[:, :, 1].clone()
target[:, :, 0] = cos_a.unsqueeze(1) * Ux - sin_a.unsqueeze(1) * Uy
target[:, :, 1] = sin_a.unsqueeze(1) * Ux + cos_a.unsqueeze(1) * Uy
```

4. **At inference, rotate velocity predictions BACK by +AoA:**
```python
Ux_pred = pred[:, :, 0].clone()
Uy_pred = pred[:, :, 1].clone()
pred[:, :, 0] = cos_a.unsqueeze(1) * Ux_pred + sin_a.unsqueeze(1) * Uy_pred
pred[:, :, 1] = -sin_a.unsqueeze(1) * Ux_pred + cos_a.unsqueeze(1) * Uy_pred
```

5. **DO NOT rotate:**
   - DSDF features (distance to surface — geometry, not flow)
   - TE coordinate frame features (already chord-relative)
   - Wake deficit features (gap-normalized, already invariant)
   - Pressure targets (scalar, rotation-invariant)

6. **Add flag** `--flowdir_norm`. No input_dim change — same channels, different values.

7. **Critical check:** Verify channel indices 0, 1 are (x, y) in the input tensor. Read `prepare_multi.py` carefully. If coordinates are at different indices, adjust accordingly.

### Training command (seed 42)
```bash
cd cfd_tandemfoil && python train.py --agent askeladd --seed 42 \
  --wandb_name "askeladd/flowdir-norm-s42" \
  --wandb_group "flowdir-anisotropic-norm" \
  --flowdir_norm \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

Seed 73: same with `--seed 73` and adjusted `--wandb_name`.

### What to report
1. 2-seed average for p_in, p_oodc, p_tan, p_re
2. W&B run IDs
3. Whether OOD metrics (p_oodc, p_re) specifically improved

## Baseline (PR #2251, 2-seed avg)

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.891** | < 11.89 |
| **p_oodc** | **7.561** | < 7.56 |
| **p_tan** | **28.118** | < 28.12 |
| p_re | 6.364 | < 6.36 |

W&B baseline: 7jix2jkg (s42), epkfhxfl (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Flow-Direction Normalization (2-seed average)

| Metric | Baseline (PR #2251) | flowdir_norm | Delta |
|--------|-------------------|-------------|-------|
| **p_in** | **11.891** | **12.630** | **+6.2% worse** |
| **p_oodc** | **7.561** | **7.832** | **+3.6% worse** |
| **p_tan** | **28.118** | **29.197** | **+3.8% worse** |
| p_re | 6.364 | 6.407 | +0.7% worse |

### Per-seed breakdown

| Metric | S42 | S73 |
|--------|-----|-----|
| p_in | 12.735 | 12.524 |
| p_oodc | 7.979 | 7.685 |
| p_tan | 29.324 | 29.070 |
| p_re | 6.487 | 6.326 |

### W&B runs
- S42: `vjao8dvz` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/vjao8dvz
- S73: `qbhm9itk` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/qbhm9itk

### Training command
```bash
cd cfd_tandemfoil && python train.py --agent askeladd --seed <42|73> \
  --wandb_name "askeladd/flowdir-norm-s<seed>" --wandb_group "flowdir-anisotropic-norm" \
  --flowdir_norm \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### Peak memory: ~68 GB (both seeds)
### Epochs: 149 (both seeds, 180-min timeout)

### What happened

**Clear negative result.** Rotating input coordinates and velocity targets by -AoA to align with the streamwise direction made all metrics worse across both seeds, with p_in degrading the most (+6.2%).

**Why it failed:**

1. **Statistics mismatch with precomputed normalization.** The input standardization stats (`x_mean`, `x_std`) were computed on un-rotated data. After rotating (x,y) by -AoA, the coordinate statistics shift slightly. While each individual rotation is small (±4° → cos≈0.998), the systematic bias from rotating ALL samples accumulates through the normalization pipeline, creating a distribution shift that the model must absorb.

2. **Redundancy with existing `aoa_perturb` augmentation.** The baseline already uses `aoa_perturb` which adds ±1° random rotation noise. This provides rotational robustness WITHOUT the statistics mismatch. Adding flowdir_norm on top doesn't add new information — it just adds a systematic offset.

3. **DSDF inconsistency.** We rotated (x,y) coordinates but NOT the DSDF gradient vectors (as instructed), creating a frame mismatch: coordinates are in the flow-aligned frame but DSDF gradients point in the original frame. While this inconsistency is small at ±4°, it likely confuses the model's spatial reasoning.

4. **Residual prediction freestream simplification didn't help.** In the rotated frame, freestream becomes a clean (1, 0, 0) vector instead of (cos(AoA), sin(AoA), 0). While algebraically elegant, the model was already well-adapted to predicting residuals relative to AoA-dependent freestream.

### Suggested follow-ups

- **Do NOT pursue further rotational canonicalization** — this confirms the SE(2) canonicalization result (#2270, also failed). The baseline's normalization pipeline and aoa_perturb augmentation are already sufficient for handling AoA variation.
- Instead, focus on improving the model's representation capacity or loss functions rather than input transformations.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2292 because ADVISOR: Closing — all metrics worse: p_in +7.6%, p_oodc +2.5%, p_tan +4.7% vs current baseline. Good analysis: (1) statistics mismatch — input standardization stats computed on un-rotated data, (2) DSDF frame inconsistency between rotated coords and un-rotated DSDF gradients, (3) redundant with existing aoa_perturb augmentation which already provides rotational robustness. Matches SE(2) canonicalization failure (#2271) — coordinate rotation without full feature transform creates inconsistencies. Flow-direction normalization added to DO NOT REVISIT. W&B: vjao8dvz (s42), qbhm9itk (s73)..
